Session Bin Test Fixes

CMCDragonkai opened the pull request:
### Description

Fixing sessions bin tests.

### Tasks

1. [x] Should be deleting the parallel tests with a new issue
2. [x] Remove nested describes
3. [x] Remove `meta?: Metadata` and replace it with: `auth` https://github.com/MatrixAI/js-polykey/pull/296#issuecomment-991845531
4. [x] Remove unnecessary `process.stdout.write` calls https://github.com/MatrixAI/js-polykey/pull/296#issuecomment-991847499
5. [x] Move `process*` options into before the `try` block that establishes the `PolykeyClient`.
6. [x] Fix up all other tests that has been affected by the changes in this PR

### Final checklist

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

tegefaulkes committed:
Fixing tests/bin/sessions, removed nested describes and concurrent test, using pkAgent

CMCDragonkai commented on tests/bin/sessions.test.ts:
```diff
@@ -163,7 +163,11 @@ describe('Session Token Refreshing', () => {
     );
 
-    test('Parallel processes should not refresh the session token', async () => {
+    // Currently requests to refresh the session token will not be dropped if
```
@joshuakarp @emmacasolin please look into creating a new issue for this.

emmacasolin replied:
Issue created for this here: https://github.com/MatrixAI/js-polykey/issues/303

CMCDragonkai commented on tests/bin/sessions.test.ts:
```diff
@@ -1,215 +1,322 @@
-describe('Session Token Refreshing', () => {
-  const logger = new Logger('pkStdio Test', LogLevel.WARN, [
+describe('CLI Sessions', () => {
+  const logger = new Logger('status test', LogLevel.WARN, [
```
Should be `sessions test`.

joshuakarp replied:
Fixed

CMCDragonkai commented on tests/bin/sessions.test.ts:
```diff
@@ -1,215 +1,322 @@
+  let sessionTokenPath: string;
   beforeAll(async () => {
-    // This handles the expensive setting up of the polykey agent.
+    sessionTokenPath = path.join(global.binAgentDir, config.defaults.tokenBase);
```
This doesn't need to be done in `beforeAll`. This can be done statically at the beginning of `describe`.

joshuakarp replied:
Fixed

CMCDragonkai commented on tests/bin/sessions.test.ts:
```diff
@@ -1,215 +1,322 @@
-      logger: logger,
+    // Invalidate all active sessions
+    await testBinUtils.pkStdio(
+      ['agent', 'lock'],
```
Forgot that this uses `agent lock`. If so then we can put these tests into the `tests/bin/agent` directory.

Also you shouldn't only do this test after first establishing a session. Unless the assumption is that you need to ensure that the session token is destroyed before running any test. In that case you should put it into `beforeEach`.

CMCDragonkai replied:
Oh I just realised this _is_ in `beforeEach`nvm.

joshuakarp replied:
Will move it over to `tests/bin/agent`. Test file still called `sessions.test.ts` I'm assuming?

CMCDragonkai replied:
Yea.

CMCDragonkai commented on tests/bin/sessions.test.ts:
```diff
@@ -1,215 +1,322 @@
+    // Run command to set token
+    let exitCode, stdout;
+    ({ exitCode, stdout } = await testBinUtils.pkStdio(
+      ['agent', 'status', '--format', 'json', '--verbose'],
```
Won't need `--format json` for this test.

joshuakarp replied:
I've removed this for all of the tests in this file - I don't believe it's necessary for any of them, but feel free to correct me @emmacasolin

joshuakarp replied:
Ignore this. Have re-added `--format json` for all tests that have an expectation on the `status` value in stdout. See https://github.com/MatrixAI/js-polykey/pull/296#discussion_r765307682

joshuakarp commented on tests/bin/sessions.test.ts:
```diff
@@ -1,215 +1,321 @@
+    ));
+    expect(exitCode).toBe(0);
+    expect(stdout).toContain('LIVE');
+    await sleep(1100);
```
Is there an alternative to doing a `sleep` here? Can we `poll` something instead?

joshuakarp replied:
Seems like there's usage of this throughout the tests in this file.

CMCDragonkai replied:
This should not be necessary. Perhaps it's waiting for expiration? We should test that separately.

CMCDragonkai replied:
Rather than waiting for expiration here. Just use the Session class and delete the token. Or just use `fs.promises.rm`.

joshuakarp replied:
Resolved by using `pkExec`now.

joshuakarp commented on tests/bin/sessions.test.ts:
```diff
@@ -1,215 +1,321 @@
+      global.binAgentDir,
+    );
+    await sleep(1100);
+    const token1 = await session.readToken();
```
Perhaps instead, for this instance, we can `poll` on `readToken`?

joshuakarp replied:
Unless `readToken` is expected to always be defined. My assumption here was that the sleep was here such that we give it time for the token to be created (i.e. `readToken` transitions from returning `undefined` to some returned token after `pk agent unlock`).

joshuakarp replied:
If we expect a change in the token, you could also just store the initial token, and `poll` until the token is different.

joshuakarp replied:
Gonna play with this.

CMCDragonkai replied:
No need to poll. Just use `status.waitFor`. See how in other tests we do this.

CMCDragonkai replied:
Oh sorry this is token...

joshuakarp replied:
Resolved by using `pkExec` now.

joshuakarp commented on tests/bin/sessions.test.ts:
```diff
@@ -1,215 +1,321 @@
+    expect(exitCode).not.toBe(0);
+    await sleep(1100);
+    // Write to session file to ensure it's unlocked
+    await session.writeToken('abc' as SessionToken);
```
Should also add an attempt to do `writeToken` at the beginning of this test, to assert that the session file is initially locked.

joshuakarp replied:
Might be misunderstanding `session` here and the test description, but I'm able to `writeFile` before the command is run with no issues (i.e. the session file is unlocked even before the command is run):

```ts
  test.only('failing processes should unlock the session file', async () => {
    const session = await Session.createSession({
      sessionTokenPath,
      fs,
      logger,
    });
    await session.writeToken('newToken' as SessionToken); // this works - and the session token becomes "newToken")
    // Run command that will fail
    const { exitCode } = await testBinUtils.pkStdio(
      ['identities', 'search', 'InvalidProvider'],
      {
        PK_NODE_PATH: global.binAgentDir,
        PK_PASSWORD: global.binAgentPassword,
      },
      global.binAgentDir,
    );
    expect(exitCode).not.toBe(0);
    await sleep(1100);
    // Write to session file to ensure it's unlocked
    await session.writeToken('abc' as SessionToken);
    const token = await session.readToken();
    expect(token).toEqual('abc');
  });
```

CMCDragonkai replied:
What do you mean `writeFile`? Yea well you can write some random crap into the token. A normal `fs.promises.writeFile` will work. The token should then be incorrect.

emmacasolin replied:
Sorry the description is poorly worded there - the session file is locked during (any) process execution so it's testing that if a command throws an error it will unlock the session file when it exits rather than leaving it locked so that future commands are able to write to the session file. This test may not be needed anymore actually, since locking behaviour was changed in the sessions review a few weeks ago.

joshuakarp replied:
Ah, I understand now. So after these locking behaviour changes, is the session file still locking during the process execution? If so, this test seems fine to me, and I'll just reword the description. However, if it isn't locked anymore, then we can remove the test entirely.

joshuakarp replied:
From @emmacasolin :

> I don’t believe so - there are separate read and write locks now so the session file shouldn’t be locked for the entire duration of the command’s execution, only when the file is being read/written to.

In that case, my opinion is this test can be removed entirely.

emmacasolin replied:
Done

joshuakarp commented on tests/bin/sessions.test.ts:
```diff
@@ -1,215 +1,321 @@
+    expect(exitCode).not.toBe(0);
+    // Prompted for password 2 times
+    expect(mockedPrompts.mock.calls.length).toBe(2);
+    mockedPrompts.mockClear();
```
Potentially a misunderstanding on my part, but what's this test achieving? As far as I understand, we provide some incorrect call, and:
1. Are prompted for a password, with which we give an incorrect password
2. Are prompted again for a password, with which we give the correct password (so no more password prompting would occur)
3. The call goes through, but it's an incorrect call, so throws some kind of error

However, the test description says "will not prompt for password reauthentication on generic error" - why would it prompt for a password reauthentication, given that we gave the correct password in step 2?

CMCDragonkai replied:
Yes this is right. It should prompt for password on generic error. So a correct test would be to provide the password on the right call in the beginning and expect an error. But how are you simulating an error here? Through deliberate mocking?

CMCDragonkai replied:
This might already be well tested in the `utils.retryAuth.test.ts`.

emmacasolin replied:
Yeah the tests here are pretty much just copied and pasted from `utils.retryAuth.test.ts`. I added them after what you said in our meeting yesterday.

emmacasolin replied:
> But how are you simulating an error here? Through deliberate mocking?

The error is thrown by `identities search` with an invalid provider. I only mocked the prompts library for these tests. So it's just checking that the error will be thrown as expected rather than catching it and prompting for the password again (as if it were an authentication error).

CMCDragonkai replied:
I've removed this test for now. It's too brittle. And the behaviour is well tested by `utils.retryAuth.test.ts`.

CMCDragonkai commented on tests/bin/sessions.test.ts:
```diff
@@ -1,215 +1,321 @@
+      },
+      global.binAgentDir,
+    );
+    await sleep(1100);
```
This should not be necessary. It is guaranteed that once the command is done, the session token is changed.

joshuakarp replied:
I've been playing around with this test (`processes should refresh the session token`), trying to remove all of the `sleep` calls.

Removing all the `sleep` calls from the test gives the following:

```ts
  test.only('processes should refresh the session token', async () => {
    const session = await Session.createSession({
      sessionTokenPath,
      fs,
      logger,
    });
    const initialToken = await session.readToken();
    logger.warn('initialToken: ' + initialToken); // TOKEN: undefined
    // Generate new token
    await testBinUtils.pkStdio(
      ['agent', 'unlock'],
      {
        PK_NODE_PATH: global.binAgentDir,
        PK_PASSWORD: global.binAgentPassword,
      },
      global.binAgentDir,
    );

    // await sleep(1100);

    const token1 = await session.readToken();
    logger.warn('token1: ' + token1); // TOKEN: ...ZV0sI1MeZ0HuSco
    // New command should refresh token
    const { exitCode, stdout } = await testBinUtils.pkStdio(
      ['agent', 'status', '--verbose'],
      {
        PK_NODE_PATH: global.binAgentDir,
      },
      global.binAgentDir,
    );
    expect(exitCode).toBe(0);
    expect(stdout).toContain('LIVE');

    // await sleep(1100);

    const token2 = await session.readToken();
    logger.warn('token2: ' + token2); // TOKEN: ...ZV0sI1MeZ0HuSco
    expect(token1).not.toBe(token2);
  });
```

Strangely, re-adding the `sleep` call immediately after the unlock:

```ts
  test.only('processes should refresh the session token', async () => {
    const session = await Session.createSession({
      sessionTokenPath,
      fs,
      logger,
    });
    const initialToken = await session.readToken();
    logger.warn('initialToken: ' + initialToken); // TOKEN: undefined
    // Generate new token
    await testBinUtils.pkStdio(
      ['agent', 'unlock'],
      {
        PK_NODE_PATH: global.binAgentDir,
        PK_PASSWORD: global.binAgentPassword,
      },
      global.binAgentDir,
    );

    await sleep(1100); // uncommented here

    const token1 = await session.readToken();
    logger.warn('token1: ' + token1); // TOKEN: ...ZV0sI1MeZ0HuSco
    // New command should refresh token
    const { exitCode, stdout } = await testBinUtils.pkStdio(
      ['agent', 'status', '--verbose'],
      {
        PK_NODE_PATH: global.binAgentDir,
      },
      global.binAgentDir,
    );
    expect(exitCode).toBe(0);
    expect(stdout).toContain('LIVE');

    // await sleep(1100);

    const token2 = await session.readToken();
    logger.warn('token2: ' + token2); // TOKEN: ...wM1Bw-NzXdP_880
    expect(token1).not.toBe(token2);
  });
```

That is, despite having no timeout added between the read of `token1` and `token2`, the tokens successfully change after having a timeout immediately after the unlock. This doesn't make a huge amount of sense to me.

joshuakarp replied:
Running the second version of the test above produces a test execution time of 1600ms according to jest output.

@CMCDragonkai suggested trying out `pkExec` - his thinking was the write to the session write is asynchronous, and may not be being awaited in the session interceptor. Using `pkExec` would wait until the process terminates, which naturally would also require the session token to have been changed.

Doing so does correctly provide the expected change in the session token. However, at least on my machine, this produces a significant increase in test time. i.e. running the following:

```ts
  test.only('processes should refresh the session token', async () => {
    const session = await Session.createSession({
      sessionTokenPath,
      fs,
      logger,
    });
    const initialToken = await session.readToken();
    logger.warn('initialToken: ' + initialToken);
    // Generate new token
    await testBinUtils.pkExec(
      ['agent', 'unlock'],
      {
        PK_NODE_PATH: global.binAgentDir,
        PK_PASSWORD: global.binAgentPassword,
      },
      global.binAgentDir,
    );

    // await sleep(1100);

    const token1 = await session.readToken();
    logger.warn('token1: ' + token1);
    // New command should refresh token
    const { exitCode, stdout } = await testBinUtils.pkExec(
      ['agent', 'status', '--verbose'],
      {
        PK_NODE_PATH: global.binAgentDir,
      },
      global.binAgentDir,
    );
    expect(exitCode).toBe(0);
    expect(stdout).toContain('LIVE');

    // await sleep(1100);

    const token2 = await session.readToken();
    logger.warn('token2: ' + token2);
    expect(token1).not.toBe(token2);
  });
```

Gives a test execution time of 11300ms according to jest:

```
 PASS  tests/bin/sessions.test.ts (26.747 s)
  CLI Sessions
    ✓ processes should refresh the session token (11369 ms)
    ○ skipped processes should store session token in session file
    ○ skipped serial processes should both refresh the session token
    ○ skipped failing processes should unlock the session file
    ○ skipped agent lock should remove the token from the client and delete the token
    ○ skipped cause old sessions to fail when locking all sessions
    ○ skipped unattended calls with invalid auth should fail
    ○ skipped password can be used to authenticate attended calls
    ○ skipped re-prompt for password if unable to authenticate call
    ○ skipped will not prompt for password reauthentication on generic error

Test Suites: 1 passed, 1 total
Tests:       9 skipped, 1 passed, 10 total
Snapshots:   0 total
Time:        26.8 s
Ran all test suites matching /tests\/bin\/sessions.test.ts/i.
GLOBAL TEARDOWN
Destroying global.binAgentDir: /run/user/1000/polykey-test-bin

```

emmacasolin replied:
The sleeps were added after I kept getting random test failures all over the place - there are likely some that are unecessary though.

joshuakarp replied:
@CMCDragonkai suggested to try this out https://www.npmjs.com/package/typescript-cached-transpile

Huge improvement when using `pkExec` now:

```ts
  test.only('processes should refresh the session token', async () => {
    const session = await Session.createSession({
      sessionTokenPath,
      fs,
      logger,
    });
    const initialToken = await session.readToken();
    logger.warn('initialToken: ' + initialToken);
    // Generate new token
    await testBinUtils.pkExec(
      ['agent', 'unlock'],
      {
        PK_NODE_PATH: global.binAgentDir,
        PK_PASSWORD: global.binAgentPassword,
        TS_NODE_TRANSPILE_ONLY: 'true',
        TS_NODE_COMPILER: 'typescript-cached-transpile',
      },
      global.binAgentDir,
    );

    // await sleep(1100);

    const token1 = await session.readToken();
    logger.warn('token1: ' + token1);
    // New command should refresh token
    const { exitCode, stdout } = await testBinUtils.pkExec(
      ['agent', 'status', '--verbose'],
      {
        PK_NODE_PATH: global.binAgentDir,
        TS_NODE_TRANSPILE_ONLY: 'true',
        TS_NODE_COMPILER: 'typescript-cached-transpile',
      },
      global.binAgentDir,
    );
    expect(exitCode).toBe(0);
    expect(stdout).toContain('LIVE');

    // await sleep(1100);

    const token2 = await session.readToken();
    logger.warn('token2: ' + token2);
    expect(token1).not.toBe(token2);
  });
```

```
 PASS  tests/bin/sessions.test.ts (14.369 s)
  CLI Sessions
    ✓ processes should refresh the session token (2401 ms)
    ○ skipped processes should store session token in session file
    ○ skipped processes should refresh the session token
    ○ skipped serial processes should both refresh the session token
    ○ skipped failing processes should unlock the session file
    ○ skipped agent lock should remove the token from the client and delete the token
    ○ skipped cause old sessions to fail when locking all sessions
    ○ skipped unattended calls with invalid auth should fail
    ○ skipped password can be used to authenticate attended calls
    ○ skipped re-prompt for password if unable to authenticate call
    ○ skipped will not prompt for password reauthentication on generic error

Test Suites: 1 passed, 1 total
Tests:       10 skipped, 1 passed, 11 total
Snapshots:   0 total
Time:        14.421 s
Ran all test suites matching /tests\/bin\/sessions.test.ts/i.
GLOBAL TEARDOWN
Destroying global.binAgentDir: /run/user/1000/polykey-test-bin
```

The first run of the test was 6000ms with these env variables set, but every successive one after was ~2400ms.

joshuakarp replied:
Hmm... I thought this problem was resolved with usage of `pkExec`, but seems like we still get random failures where the session tokens are still the same after execution of the command (solely in this `processes should refresh the session token` test).

CMCDragonkai replied:
Should be resolved with our new `FlowCountInterceptor`.

CMCDragonkai commented on tests/bin/sessions.test.ts:
```diff
@@ -1,215 +1,321 @@
+      global.binAgentDir,
+    );
+    expect(exitCode).toBe(0);
+    expect(stdout).toContain('LIVE');
```
@joshuakarp please replace these with JSON.parse and match object on status.

joshuakarp replied:
So any usage of `pk agent status` should be using `--format json` and parsing the output then? Your original comment said to get rid of this - https://github.com/MatrixAI/js-polykey/pull/296#discussion_r764853205

joshuakarp replied:
Fixed https://github.com/MatrixAI/js-polykey/pull/296/commits/5ea743728d69b42f3a324d4fa6d093a7ce72faeb

CMCDragonkai commented on tests/bin/sessions.test.ts:
```diff
@@ -1,215 +1,321 @@
+    );
+    expect(exitCode).toBe(0);
+    expect(stdout).toContain('LIVE');
+    await sleep(1100);
```
There is no need to sleep. The token should always be refreshed by the time the command finishes. Token refreshing is done on the client side and therefore it is not like `pk agent stop`.

joshuakarp replied:
All sleeps removed https://github.com/MatrixAI/js-polykey/pull/296#discussion_r765381558

CMCDragonkai commented on tests/bin/sessions.test.ts:
```diff
@@ -1,215 +1,321 @@
+    ));
+    expect(exitCode).toBe(0);
+    expect(stdout).toContain('LIVE');
+    await sleep(1100);
```
Remove all of these sleeps.

joshuakarp replied:
All sleeps have been removed in favour of using `pkExec` (to enforce the asynchronous session token write completes at conclusion of command) with the newly added typescript-cached-transpile (to improve execution time of the tests).

There's one residual sleep in `failing processes should unlock the session file`, but this test will potentially be removed, as per https://github.com/MatrixAI/js-polykey/pull/296#discussion_r765357879

CMCDragonkai commented on tests/bin/sessions.test.ts:
```diff
@@ -1,215 +1,321 @@
+  test('cause old sessions to fail when locking all sessions', async () => {
+    // Generate new token
+    await testBinUtils.pkStdio(
+      ['agent', 'unlock'],
```
You might as well add in a test for `lockall` too then at this point.

Later we can refactor some locking tests to be in `lock.test.ts`and `unlock.test.ts`and `lockall.test.ts`. Then `sessions.test.ts` will be what is left over.

CMCDragonkai replied:
Specifically I wanted `sessions.test.ts` to be precisely just the CARL loop.

emmacasolin replied:
All of the locking tests have been moved to new files in `bin/agent/unlock.test.ts`, `bin/agent/lock.test.ts`, and `bin/agent/lockall.test.ts`.

joshuakarp committed:
Adding typescript-cache-transpile for faster test execution with processes (usage with pkExec, pkSpawn, pkExpect)

CMCDragonkai committed:
Session fixes: correct authentication usage for CLI commands, exceptions thrown, deletion of session token on lock

CMCDragonkai committed:
Integration FlowCountInterceptor to GRPCClient, enabling pkStdio

CMCDragonkai committed:
Ensure proper return for callback for `agentStop` and `sessionsUnlock` and `sessionsLockAll`

emmacasolin committed:
Adding returns to all GRPC handlers
Parsing new password seperately from existing password
Agent lock, unlock and lockall tests

CMCDragonkai commented on src/bin/keys/CommandRenew.ts:
```diff
@@ -51,10 +52,19 @@ class CommandRenew extends CommandPolykey {
+        let password: string | undefined;
+        try {
+          password = (
+            await this.fs.promises.readFile(passwordPath, 'utf-8')
```
This is meant to be a new option in `binOptions`. It's not supposed to just read a file from the password path.

CMCDragonkai replied:
We can make it a mandatory option. You have to set it. Alternatively we can also prompt for it.

CMCDragonkai replied:
Done with `binOptions.passwordNewFile` and `processNewPassword`.

emmacasolin replied:
I did it that way because there's a similar parameter for some of the keys commands (e.g. `encrypt` and `decrypt`) that also take a file as a parameter and process it like that. Should this also be made into a bin option?

CMCDragonkai replied:
No that's different. Passwords are special cases.

CMCDragonkai committed:
`pk keys password`, `pk keys renew` and `pk keys reset` all take `--password-new-file` now

CMCDragonkai committed:
`pk agent start` and `pk bootstrap` are also using `processNewPassword`, `processPassword` is only used by `pk agent start`

CMCDragonkai committed:
Added test for stopping agent while unauthenticated, fixed `pk agent start` when starting with non-existing node path

CMCDragonkai committed:
Reading `PK_PASSWORD` should be in `processNewPassword`

CMCDragonkai committed:
Fixing password env

emmacasolin committed:
sessions tests passing

CMCDragonkai committed:
Close the testLockFile for `pkAgent` after unlocking

CMCDragonkai committed:
Fixing up `pk agent lock` tests

emmacasolin committed:
test fixes for keys + sessions

CMCDragonkai committed:
Fixing `pk agent lock` and `pk agent lockall` tests

CMCDragonkai committed:
Fixed tests/bin/agent concurrent testing

emmacasolin committed:
nonces comment on sessions tests

CMCDragonkai commented on tests/client/rpcSessions.test.ts:
```diff
@@ -101,81 +85,27 @@ describe('Sessions client service', () => {
-    // Locking the session.
-    const sessionLockAll = grpcUtils.promisifyUnaryCall<utilsPB.EchoMessage>(
+  test('can lock all sessions', async () => {
+    const requestJWT = grpcUtils.promisifyUnaryCall<utilsPB.EmptyMessage>(
```
Didn't we stop using the `JWT` terminology when dealing with sessions?

emmacasolin replied:
Yeah that sounds familiar, I just left it as it was but I'll give it a better name now

CMCDragonkai committed:
Agent and sessions tests working

emmacasolin committed:
test fixes from status command implementation

emmacasolin committed:
more meaningful naming

emmacasolin committed:
replacing meta?: Metadata with auth

emmacasolin committed:
missed metas in vaults commands

CMCDragonkai committed:
Moved `PolykeyClient.test.ts` out of `tests/client` to `tests`

emmacasolin committed:
moving preprocessing outside of try blocks

emmacasolin committed:
removing unnecessary writes to stdout

CMCDragonkai committed:
Setup global root keypair for `tests/PolykeyClient.test.ts`

CMCDragonkai committed:
Linted

CMCDragonkai committed:
Added timeouts to bootstrapping
